src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,167 @@
+#include "base_alloc_internal.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+// minimum size of a single pool of the base allocator,
+#define MINIMUM_POOL_SIZE (ba_os_get_page_size())
+
+// minimum number of chunks used to calculate the size of pools
+#define MINIMUM_CHUNK_COUNT (128)
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct ba_mem_chunk_t ba_mem_chunk_t;
+
+// memory chunk of size 'chunk_size'
+struct ba_mem_chunk_t {
+    ba_mem_chunk_t *next;
+};
+
+struct umf_ba_pool_meta_t {
+    // address of the beginning of this pool (allocated region)
+    umf_ba_pool_t *pool_addr;
+    // address of the beginning of the next pool (a list of allocated pools to be freed in umf_ba_destroy())
+    umf_ba_pool_t *next_pool;
+
+    // the following fields are set and used only in the main (the first) pool
+    size_t pool_size; // size of each pool (argument of each ba_os_alloc() call)
+    size_t pool_usable_size;   // usable size of each pool
+    size_t chunk_size;         // size of all memory chunks in this pool
+    os_mutex_t *free_lock;     // lock of free_list
+    ba_mem_chunk_t *free_list; // list of free chunks
+};
+
+struct umf_ba_pool_t {
+    struct umf_ba_pool_meta_t metadata;
+    char data[];
+};
+
+// ba_divide_memory_into_chunks - divide given memory into chunks of chunk_size and add them to the free_list
+static void ba_divide_memory_into_chunks(umf_ba_pool_t *pool, void *ptr,
+                                         size_t size) {
+    assert(pool->metadata.free_list == NULL);
+    assert(size > pool->metadata.chunk_size);
+
+    char *pool_memory = ptr;
+    size_t size_left = size;
+
+    ba_mem_chunk_t *current_chunk = (ba_mem_chunk_t *)pool_memory;
+    ba_mem_chunk_t *prev_chunk = current_chunk;
+
+    while (size_left >= pool->metadata.chunk_size) {
+        current_chunk = (ba_mem_chunk_t *)pool_memory;
+        prev_chunk->next = current_chunk;
+
+        pool_memory += pool->metadata.chunk_size;
+        size_left -= pool->metadata.chunk_size;
+        prev_chunk = current_chunk;
+    }
+
+    current_chunk->next = NULL;
+    pool->metadata.free_list = ptr; // address of the first chunk
+}
+
+umf_ba_pool_t *umf_ba_create(size_t size) {
+    size_t chunk_size = align_size(size, MEMORY_ALIGNMENT);
+    size_t mutex_size = align_size(util_mutex_get_size(), MEMORY_ALIGNMENT);
+    size_t metadata_size = offsetof(struct umf_ba_pool_t, data);
+    size_t pool_size =
+        metadata_size + mutex_size + (MINIMUM_CHUNK_COUNT * chunk_size);
+    if (pool_size < MINIMUM_POOL_SIZE) {
+        pool_size = MINIMUM_POOL_SIZE;
+    }
+
+    pool_size = align_size(pool_size, ba_os_get_page_size());
+
+    umf_ba_pool_t *pool = (umf_ba_pool_t *)ba_os_alloc(pool_size);
+    if (!pool) {
+        return NULL;
+    }
+
+    pool->metadata.pool_addr = pool;
+    pool->metadata.pool_size = pool_size;
+    pool->metadata.pool_usable_size = pool_size - metadata_size;
```

Comment:
if you create functions calc_pool_size(size) and calc_metadata_size(size) you don't need to keep these values here

Reply:
`pool_size` is calculated once and stored only in one place now. This is simpler and faster.